pkg/virt-controller/watch/replicaset.go

```diff
@@ -0,0 +1,372 @@
+		for i := diff; i < 0; i++ {
+			go func() {
+				defer wg.Done()
+				vm := virtv1.NewVMReferenceFromNameWithNS(rs.ObjectMeta.Namespace, "")
+				vm.ObjectMeta = rs.Spec.Template.ObjectMeta
+				vm.ObjectMeta.Name = ""
+				vm.ObjectMeta.GenerateName = basename
+				vm.Spec = rs.Spec.Template.Spec
+				// TODO check if vm labels exist, and when make sure that they match. For now just override them
+				vm.ObjectMeta.Labels = rs.Spec.Template.ObjectMeta.Labels
+				_, err := c.clientset.VM(rs.ObjectMeta.Namespace).Create(vm)
+				if err != nil {
+					errChan <- err
+				}
+			}()
+		}
+	}
+	wg.Wait()
+
+	select {
+	case err := <-errChan:
+		// Only return the first error which occurred, the others will most likely be equal errors
+		return err
+	default:
+	}
+	return nil
+}
+
+// filterActiveVMs takes a list of VMs and returns all VMs which are not in a final state
+// Note that vms which have a deletion timestamp set, are still treated as active.
+// This is a difference to Pod ReplicaSets
+func (c *VMReplicaSet) filterActiveVMs(vms []virtv1.VM) []virtv1.VM {
+	filtered := []virtv1.VM{}
+	for _, vm := range vms {
+		if !vm.IsFinal() {
+			filtered = append(filtered, vm)
+		}
+	}
+	return filtered
+}
+
+// filterMatchingVMs takes a selector and a list of VMs. If the VM labels match the selector it is added to the filtered collection.
+// Returns the list of all VMs which match the selector
+func (c *VMReplicaSet) filterMatchingVMs(selector labels.Selector, vms []virtv1.VM) []virtv1.VM {
+	//TODO take owner reference into account
+	filtered := []virtv1.VM{}
+	for _, vm := range vms {
+		if selector.Matches(labels.Set(vm.ObjectMeta.Labels)) {
+			filtered = append(filtered, vm)
+		}
+	}
+	return filtered
+}
+
+// listVMsFromNamespace takes a namespace and returns all VMs from the VM cache which run in this namespace
+func (c *VMReplicaSet) listVMsFromNamespace(namespace string) ([]virtv1.VM, error) {
+	objs, err := c.vmInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+	vms := []virtv1.VM{}
+	for _, obj := range objs {
+		vms = append(vms, *obj.(*virtv1.VM))
+	}
+	return vms, nil
+}
+
+// listControllerFromNamespace takes a namespace and returns all VMReplicaSets from the ReplicaSet cache which run in this namespace
+func (c *VMReplicaSet) listControllerFromNamespace(namespace string) ([]virtv1.VirtualMachineReplicaSet, error) {
+	objs, err := c.vmRSInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+	replicaSets := []virtv1.VirtualMachineReplicaSet{}
+	for _, obj := range objs {
+		rs := obj.(*virtv1.VirtualMachineReplicaSet)
+		replicaSets = append(replicaSets, *rs)
+	}
+	return replicaSets, nil
+}
+
+// vmChangeFunc checks if the supplied VM matches a replica set controller in it's namespace
+// and wakes the first controller which matches the VM labels.
+func (c *VMReplicaSet) vmChangeFunc(obj interface{}) {
```

Comment:
One common pattern people use when replicating resources, whether it be containers or VMs, is the concept of isolating a resource from the group controller.

If someone manually changed the labels on a VM to isolate it from the replication controller, and then decides to terminate the resource, I don't believe the replication controller would ever get triggered to resize.

Basically, if someone changes a label on a VM and then deletes the VM, the VM replica set would be permanently at n-1 VMs until something else triggered the replica set to get processed.

Reply:
The ReplicaSet in k8s reacts to relabeling of Pods. If the labels do not match anymore, it either gets adopted by another matching ReplicaSet, or it stays in an orphaned state. The original ReplicaSet is woken up and will immediately create a new Pod.

Reply:
Actually, looks like I was wrong.

That [comment](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L272) indicates that it would wake up both controllers. Might be an artifact of the pre-controller-ref aera. However, the actual code in that function only reacts to label changes if the controller reference changes, or if the controller behind the reference was deleted. To detach a resource for sure from the controller (also prevent re-adoption by another controller), looks like one has to change labels and delete the controller ref. If only the labels are changed, it will still stick with the crating controller. If the controller reference is removed but the labels still stay the same it will be adopted again.

I will take that into account when adding the controller-ref. :+1:

Reply:
Ok, my original comment was right. The flow is as follows with controller-refs:

1) if labels changed, inform the owner only
2) the owner will try to release the object (remove the owner ref)
3) this causes another update on the resource. Now there is no owner, so all matching remaining controllers are informed
4) they race for adopting that resource

So changing a label on a Pod in the case of ReplicaSets leads to releasing the object and reclaiming it by another controlelr if one matches. Further the original ReplicaSet creates a new Pod  as a replacement for the old pod.